### PR TITLE
Default functions, extend YarnCommand and YarnFunction attributes

### PR DESCRIPTION
**Pull request by xsduan**
* **Please check if the pull request fulfills these requirements**
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] CHANGELOG.md has been updated to describe this change

* **What kind of change does this pull request introduce?**

- [ ] Bug Fix
- [x] Feature
- [ ] Something else

It seems like the function support was not fully implemented (a NotImplementedException was being thrown when adding functions) but I have taken the liberty to fully implement it, and add a few features of my own. Not sure which issues I could link. Hopefully that's not overstepping too much!

* **What is the current behavior?** 

The command support is somewhat incomplete, and no functions by default are added.

* **What is the new behavior (if this is a feature change)?**

I'm just gonna copy everything from the changelog to avoid repeating myself.

- Default functions have been added for convenience.
  - `float random()` - returns a number between 0 and 1, inclusive (proxies Unity's default prng)
  - `float random_range(float, float)` - returns a number in a given range, inclusive (proxies Unity's default prng)
  - `int dice(int)` - returns an integer in a given range, like a dice (proxies Unity's default prng)
    - For example, `dice(6) + dice(6)` to simulate two dice, or `dice(20)` for a D20 roll.
  - `int round(float)` - rounds a number using away-from-zero rounding
  - `float round_places(float, int)` - rounds a number to n digits using away-from-zero rounding
  - `int floor(float)` - floors a number (towards negative infinity)
  - `int ceil(float)` - ceilings a number (towards positive infinity)
  - `int int(float)` - truncates the number (towards zero)
  - `int inc(float | int)` - increments to the next integer
  - `int dec(float | int)` - decrements to the previous integer
  - `int decimal(float)` - gets the decimal portion of the float

- The `YarnFunction` attribute has been added.
  - Simply add it to a static function, eg

    ```cs
    [YarnFunction] // registers function under "example"
    public static int example(int param) {
      return param + 1;
    }

    [YarnFunction("custom_name")] // registers function under "custom_name"
    public static int example2(int param) {
      return param * param;
    }
    ```

- The `YarnCommand` attribute has been improved and made more robust for most use cases.
  - You can now leave the name blank to use the method name as the registration name.

    ```cs
    [YarnCommand] // like in previous example with YarnFunction.
    void example(int steps) {
      for (int i = 0; i < steps; i++) { ... }
    }

    [YarnCommand("custom_name")] // you can still provide a custom name if you want
    void example2(int steps) {
      for (int i = steps - 1; i >= 0; i--) { ... }
    }
    ```

  - It now recognizes static functions and does not attempt to use the first parameter as an instance variable anymore.

    ```cs
    [YarnCommand] // use like so: <<example>>
    static void example() => ...;

    [YarnCommand] // still as before: <<example2 objectName>>
    void example2() => ...;
    ```

  - You can also define custom getters for better performance.
  
    ```cs
    [YarnStateInjector(nameof(GetBehavior))] // if this is null, the previous behavior of using GameObject.Find will still be available
    class CustomBehavior : MonoBehaviour {
      static CustomBehavior GetBehavior(string name) {
        // e.g., it may only exist under a certain transform, or you have a custom cache...
        // or it's built from a ScriptableObject...
        return ...;
      }

      [YarnCommand] // the "this" will be as returned from GetBehavior
      void example() => Debug.Log(this);

      // special variation on getting behavior
      static CustomBehavior GetBehaviorSpecial(string name) => ...;

      [YarnCommand(Injector = nameof(GetBehaviorSpecial))]
      void example_special() => Debug.Log(this);
    }
    ```

  - You can also define custom getters for Component parameters in the same vein.
  
    ```cs
    class CustomBehavior : MonoBehaviour {
      static Animator GetAnimator(string name) => ...;

      [YarnCommand]
      void example([YarnParameter(nameof(GetAnimator))] Animator animator) => Debug.Log(animator);
    }
    ```

  - You should continue to use manual registration if you want to have "pseudo-static" dispatch, where there is an instance attached to the action, but the script should not know about it.

* **Does this pull request introduce a breaking change?**

I have tried to avoid breaking changes and designed the changes to default to the previous behavior but obviously with a change this drastic, there is bound to be an edge case or two that I missed.

However, I did not have to "fix" any existing unit tests in this area (other than a couple wording changes), so that is a promising sign.

* **Other information**:

### CODE REVIEW

**Comment by xsduan** [comment]
How low do we HAVE to test here? I got it to work on 2019 but couldn't really get it to work on 2018. My personal opinion is that not enough people are using 2018.4 anymore to justify testing it (especially once we take into account that by the time we exit beta LTS will drop).

Additionally, I did see this (which I did not modify):

```cs
static IEnumerator WaitForYieldInstruction(Delegate @theDelegate, object[] finalParametersToUse, Action onSuccessfulDispatch)
{
    yield return @theDelegate.DynamicInvoke(finalParametersToUse);
    onSuccessfulDispatch();
}
```

which does not compile on 2019. It looks like it's c# 8 which might just be one of those quirks of Unity supporting features from c# 8 randomly???

If we don't actually need to support 2018.4 then we should probably bump it up to 2019 or 2020 in the dependencies.

**Review by McJones — COMMENTED** [review]
Wow this is amazing, when you said you wanted to do this I don't know what I was expecting but this is so impressive.

I have no real comments other than a couple of little nitpicks here and there.
I would just merge this anyways but I would want to work out do we want to make it a recommendation (not requirement) for functions in Yarn to be camelCase or snake_case, but that is a 10 second change even if we decide to go down the camelCase route.

Finally as to your comment about Unity versions while not 100% locked in at this stage based on a quick poll we ran in the YS discord its looking like there is at a minimum a 10:1 difference in 2018 vs other versions.
So I reckon we will end up dropping 2018 support and I think if you go ahead and explicitly drop support for it in this PR it will be fine.
If we end up supporting 2018 we'll go ahead and make that side work.

File: Runtime/Commands/ActionManager.cs [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+                        $"the command is only defined on {behaviorType.FullName} components, but {name} doesn't have one.");
+                    return null;
+                }
+                return target;
+            };
+        }
+
+        private static bool IsInjector(Type injectorType, MethodInfo injectorFunction, Type destinationType = null)
+        {
+#if UNITY_2020_3_OR_NEWER
```
[the remark this answers is not in the document]
Comment (xsduan) [reply Runtime/Commands/ActionManager.cs:49]: Well personally I think that it's much cleaner to use the ??= operator as it shows more intent, and I don't want anyone to forget to upgrade when we drop support for 2019 in 2023 or something. 

I guess this is coming from my Typescript experience where everything can be neatly downleveled automatically. Not sure if that makes sense here, but just explaining my thought process

File: Runtime/Commands/ActionManager.cs [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+            {
+                Method = method,
+                Injector = injector,
+                Converters = converters
+            };
+        }
+
+        private static Type GetFuncType(int paramCount)
+        {
+            // this causes me physical pain
```
[the remark this answers is not in the document]
Comment (xsduan) [reply Runtime/Commands/ActionManager.cs:191]: I think iOS specifically prohibits codegen (eg MessagePack is forced to AOT because of this). Not sure if dynamically creating a delegate would count?

File: Runtime/Commands/ActionManager.cs [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+            return finalArgs;
+        }
+
+        /// <summary>
+        /// Register a function into a given library.
+        /// </summary>
+        /// <param name="library">Library instance to register.</param>
+        public static void RegisterFunctions(Library library)
+        {
+#if NET_STANDARD // c# 8+ (Unity 2022)
```
[the remark this answers is not in the document]
Comment (xsduan) [reply Runtime/Commands/ActionManager.cs:392]: `KeyValuePair<TKey, TValue>` only gains a `Deconstruct` member in .NET Standard 2.1 and I just think that's a lot nicer than `.Key` and `.Value`. Falls under the other similar issues' discussions.

File: Runtime/Commands/DefaultActions.cs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+        #region Functions
+        [YarnFunction]
+        public static float random()
+        {
+            return random_range(0, 1);
+        }
+
+        [YarnFunction]
+        public static float random_range(float minInclusive, float maxInclusive)
```
[the remark this answers is not in the document]
Comment (xsduan) [reply Runtime/Commands/DefaultActions.cs:32]: I don't have a particular preference. @desplesda tiebreaker?

[the remark this answers is not in the document]
Comment (desplesda) [reply Runtime/Commands/DefaultActions.cs:32]: I think `snake_case` is a little easier to read and type than `camelCase`. Tie: broken!

File: Runtime/Commands/ActionManager.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (desplesda) [reply Runtime/Commands/ActionManager.cs:191]: I think Tim means using something like a compile-time code generator, not runtime JIT stuff (which is very much not allowed in AOT environments).

File: Runtime/Commands/DefaultActions.cs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        }
+
+        [YarnFunction]
+        public static int round(float num)
+        {
+            return (int)round_places(num, 0);
+        }
+
+        [YarnFunction]
+        public static float round_places(float num, int places)
```
[the remark this answers is not in the document]
Comment (desplesda) [reply Runtime/Commands/DefaultActions.cs:56]: Likewise, I'm decreeing `snake_case` here. 🐍

File: Runtime/Commands/ActionManager.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (xsduan) [reply Runtime/Commands/ActionManager.cs:191]: Looked into it a little bit, it seems like support isn't 100% there for source generators (because Nuget). I think we should probably leave it for now and revisit it when 2021 goes into LTS.

File: Runtime/Commands/ActionManager.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (McJones) [reply Runtime/Commands/ActionManager.cs:49]: fair enough, it doesn't hurt I guess to leave as is anyways.

File: Runtime/Commands/ActionManager.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (McJones) [reply Runtime/Commands/ActionManager.cs:191]: agreed

File: Runtime/Commands/ActionManager.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (McJones) [reply Runtime/Commands/ActionManager.cs:392]: same, this is fine

**Comment by McJones** [comment]
I guess I have one more comment before I merge this (or I can do it if you are doing other stuff) we are going with snake_case for functions but C# still goes down the camelCase route, so can you update the `round_places` and `random_range` to be camelCase in the C# and using the attribute make it snake for Yarn?

Let me know either way, thanks again for the excellent PR.

**Comment by xsduan** [comment]
Personally I feel like since these aren't referenced elsewhere we should take advantage of the auto-naming system, but I don't really mind either way
